File: Trigger/TrigT1/TrigT1CaloMonitoring/python/GfexMonitorAlgorithm.py
#
#  Copyright (C) 2002-2022 CERN for the benefit of the ATLAS collaboration
#
def GfexMonitoringConfig(inputFlags):
    '''Function to configure LVL1 Gfex algorithm in the monitoring system.'''

    import math

    # get the component factory - used for merging the algorithm results
    from AthenaConfiguration.ComponentFactory import CompFactory
    from AthenaConfiguration.ComponentAccumulator import ComponentAccumulator
    result = ComponentAccumulator()
    
    # uncomment if you want to see all the flags
    #inputFlags.dump() # print all the configs

    # make the athena monitoring helper
    from AthenaMonitoring import AthMonitorCfgHelper
    helper = AthMonitorCfgHelper(inputFlags,'GfexMonitoringCfg')

    # get any algorithms
    GfexMonAlg = helper.addAlgorithm(CompFactory.GfexMonitorAlgorithm,'GfexMonAlg')

    # add any steering
    groupName = 'GfexMonitor' # the monitoring group name is also used for the package name
    GfexMonAlg.PackageName = groupName

    mainDir = 'L1Calo'
    trigPath = 'Gfex/'

    etabins=32
    etamin=-3.3
    etamax=3.3
    phibins=34
    phimin=-3.3
    phimax=3.3

    # add monitoring algorithm to group, with group name and main directory 
    myGroup = helper.addGroup(GfexMonAlg, groupName , mainDir)

    # define gfex histograms
    myGroup.defineHistogram('gFexLRJetTransverseEnergy;h_gFexLRJetTransverseEnergy',
                            title='gFex LRJet Transverse Energy; gFexLRJet Et [MeV]; counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=0,xmax=819200)

    myGroup.defineHistogram('gFexLRJetEta;h_gFexLRJetEta',
                            title='gFex LRJet #eta; #eta; counts',
                            type='TH1F', path=trigPath,
                            xbins=etabins,xmin=etamin,xmax=etamax)

    myGroup.defineHistogram('gFexLRJetPhi;h_gFexLRJetPhi',
                            title='gFex LRJet #phi;  #phi; counts',
                            type='TH1F', path=trigPath,
                            xbins=phibins,xmin=phimin,xmax=phimax)

    myGroup.defineHistogram('gFexRhoTransverseEnergy;h_gFexRhoTransverseEnergy',
                            title='gFex Rho Transverse Energy; gFexRho Et [MeV]; counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-300000,xmax=300000)


    myGroup.defineHistogram('gFexSRJetTransverseEnergy;h_gFexSRJetTransverseEnergy',    
                            title='gFex SRJet Transverse Energy; gFexSRJet Et [MeV]; counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-100,xmax=819200)

    myGroup.defineHistogram('gFexSRJetEta;h_gFexSRJetEta',
                            title='gFex SRJet #eta; #eta; counts',
                            type='TH1F', path=trigPath,
                            xbins=etabins,xmin=etamin,xmax=etamax)

    myGroup.defineHistogram('gFexSRJetPhi;h_gFexSRJetPhi',
                            title='gFex SRJet #phi; #phi; counts',
                            type='TH1F', path=trigPath,
                            xbins=32,xmin=-math.pi,xmax=math.pi)

    myGroup.defineHistogram('gFexType;h_gFexType',
                            title='gFex Type for all Trigger Objects; gFex Types',
                            xlabels=["gRho", "gBlockLead", "gBlockSub", "gJet"],
                            type='TH1F', path=trigPath,
                            xbins=4,xmin=0,xmax=4)

    myGroup.defineHistogram('gFexLRJetEta,gFexLRJetPhi;h_gFexLRJetEtaPhiMap_weighted',
                            title="gFexLRJet #eta vs #phi;gFexLRJet #eta;gFexLRJet #phi;gFexLRJet TransverseEnergy",
                            type='TH2F',path=trigPath,weight="gFexLRJetTransverseEnergy",
                            xbins=etabins,xmin=etamin,xmax=etamax,ybins=32,ymin=0,ymax=math.pi)

    myGroup.defineHistogram('gFexSRJetEta,gFexSRJetPhi;h_gFexSRJetEtaPhiMap_weighted',
                            title="gFexSRJet #eta vs #phi;gFexSRJet #eta;gFexSRJet #phi;gFexSRJet TransverseEnergy",
                            type='TH2F',path=trigPath,weight="gFexSRJetTransverseEnergy",
                            xbins=etabins,xmin=etamin,xmax=etamax,ybins=32,ymin=0,ymax=math.pi)
    
    myGroup.defineHistogram('gFexLRJetEta,gFexLRJetPhi;h_gFexLRJetEtaPhiMap',
                            title="gFexLRJe #eta vs #phi;gFexLRJet #eta;gFexLRJet #phi",
                            type='TH2F',path=trigPath,
                            xbins=etabins,xmin=etamin,xmax=etamax,ybins=32,ymin=0,ymax=math.pi)

    myGroup.defineHistogram('gFexSRJetEta,gFexSRJetPhi;h_gFexSRJetEtaPhiMap',
                            title="gFexSRJet #eta vs #phi;gFexSRJet #eta;gFexSRJet #phi",
                            type='TH2F',path=trigPath,
                            xbins=etabins,xmin=etamin,xmax=etamax,ybins=32,ymin=0,ymax=math.pi)

    myGroup.defineHistogram('gFexMET;h_gFexMET',
                            title='gFex MET obtained with jets without jets algorithm;MET [MeV];counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=0,xmax=600000)

    myGroup.defineHistogram('gFexSumET;h_gFexSumET',
                            title='gFex SumET obtained with jets without jets algorithm ; SumET [MeV];counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=0,xmax=5000000)

    myGroup.defineHistogram('gFexMETx;h_gFexMETx',
                            title='gFex MET x obtained with jets without jets algorithm ; MET x [MeV];counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-300000,xmax=300000)

    myGroup.defineHistogram('gFexMETy;h_gFexMETy',
                            title='gFex MET y obtained with jets without jets algorithm ; MET y [MeV];counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-300000,xmax=300000)

    myGroup.defineHistogram('gFexMHTx;h_gFexMHTx',
                            title='gFex MHT x obtained with jets without jets algorithm; MHT x [MeV];counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-300000,xmax=300000)

    myGroup.defineHistogram('gFexMHTy;h_gFexMHTy',
                            title='gFex MHT y obtained with jets without jets algorithm;MHT y [MeV];counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-300000,xmax=300000)

    myGroup.defineHistogram('gFexMSTx;h_gFexMSTx',
                            title='gFex MST x obtained with jets without jets algorithm;MST x [MeV];counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-200000,xmax=200000)

    myGroup.defineHistogram('gFexMSTy;h_gFexMSTy',
                            title='gFex MST y obtained with jets without jets algorithm;MST y [MeV];counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-200000,xmax=200000)

    myGroup.defineHistogram('gFexMETx_NoiseCut;h_gFexMETx_NoiseCut',
                            title='gFex MET x obtained with noise cut algorithm;MET x [MeV]; counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-600000,xmax=600000)

    myGroup.defineHistogram('gFexMETy_NoiseCut;h_gFexMETy_NoiseCut',
                            title='gFex MET y obtained with noise cut algorithm;MET y [MeV]; counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-600000,xmax=600000)

    myGroup.defineHistogram('gFexMETx_ComponentsRms;h_gFexMETx_ComponentsRms',
                            title='gFex MET x obtained with Rho+RMS algorithm;MET x [MeV]; counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-600000,xmax=600000)

    myGroup.defineHistogram('gFexMETy_ComponentsRms;h_gFexMETy_ComponentsRms',
                            title='gFex MET y obtained with Rho+RMS algorithm;MET y [MeV]; counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-600000,xmax=600000)

    myGroup.defineHistogram('gFexMET_NoiseCut;h_gFexMET_NoiseCut',
                            title='gFex MET [MeV] obtained with noise cut algorithm;MET [MeV]; counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-10000,xmax=1000000)

    myGroup.defineHistogram('gFexSumET_NoiseCut;h_gFexSumET_NoiseCut',
                            title='gFex SumET [MeV] obtained with noise cut algorithm;Sum ET [MeV]; counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=0,xmax=2500000)

    myGroup.defineHistogram('gFexMET_Rms;h_gFexMET_Rms',
                            title='gFex MET [MeV] obtained with Rho+RMS algorithm;MET [MeV]; counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=-100000,xmax=1000000)

    myGroup.defineHistogram('gFexSumET_Rms;h_gFexSumET_Rms',
                            title='gFex SumET [MeV] obtained with Rho+RMS algorithm;Sum ET [MeV]; counts',
                            type='TH1F', path=trigPath,
                            xbins=100,xmin=0,xmax=2000000)

    acc = helper.result()
    result.merge(acc)
    return result


if __name__=='__main__':
    # set input file and config options
    from AthenaConfiguration.AllConfigFlags import ConfigFlags
    import glob

    # MCs with SCells
    #inputs = glob.glob('/eos/user/t/thompson/ATLAS/LVL1_mon/valid1.361021.Pythia8EvtGen_A14NNPDF23LO_jetjet_JZ1W.recon.ESD.e3569_s3126_d1623_r12488/ESD.24607652._000025.pool.root.1')
    #inputs = glob.glob('/eos/atlas/atlascerngroupdisk/det-l1calo/OfflineSoftware/mc16_13TeV.361106.PowhegPythia8EvtGen_AZNLOCTEQ6L1_Zee.recon.ESD.e3601_s3126_r12406/ESD.24368002._000045.pool.root.1')
    #
    # Above MCs processed adding L1_eEMRoI
    inputs = glob.glob('/afs/cern.ch/user/t/thompson/work/public/LVL1_monbatch/run_sim/l1calo.361106.PowhegPythia8EvtGen_AZNLOCTEQ6L1_Zee.ESD.eFex_2021-05-16T2101.root')
    

    ConfigFlags.Input.Files = inputs
    ConfigFlags.Output.HISTFileName = 'ExampleMonitorOutput.root'

    ConfigFlags.lock()
    ConfigFlags.dump() # print all the configs

    from AthenaCommon.AppMgr import ServiceMgr
    ServiceMgr.Dump = False

    from AthenaConfiguration.MainServicesConfig import MainServicesCfg  
    from AthenaPoolCnvSvc.PoolReadConfig import PoolReadCfg
    cfg = MainServicesCfg(ConfigFlags)
    cfg.merge(PoolReadCfg(ConfigFlags))

    GfexMonitorCfg = GfexMonitoringConfig(ConfigFlags)
    cfg.merge(GfexMonitorCfg)

    # options - print all details of algorithms, very short summary 
    cfg.printConfig(withDetails=False, summariseProps = True)

    nevents=100
    cfg.run(nevents)

File: Trigger/TrigT1/TrigT1CaloMonitoring/src/GfexMonitorAlgorithm.cxx
/*
  Copyright (C) 2002-2022 CERN for the benefit of the ATLAS collaboration
*/

#include "GfexMonitorAlgorithm.h"

#include "CaloDetDescr/CaloDetDescrElement.h"
#include "CaloIdentifier/CaloGain.h"
#include "Identifier/Identifier.h"
#include "Identifier/HWIdentifier.h"
#include "CaloIdentifier/CaloCell_ID.h"
#include "CaloDetDescr/CaloDetDescrManager.h"
#include <istream>


GfexMonitorAlgorithm::GfexMonitorAlgorithm( const std::string& name, ISvcLocator* pSvcLocator )
  : AthMonitorAlgorithm(name,pSvcLocator)
{
}

StatusCode GfexMonitorAlgorithm::initialize() {
  ATH_MSG_DEBUG("GfexMonitorAlgorithm::initialize");
  ATH_MSG_DEBUG("Package Name "<< m_packageName);
  
  ATH_MSG_DEBUG("m_gFexLRJetContainerKey"<< m_gFexLRJetContainerKey);
  ATH_MSG_DEBUG("m_gFexSRJetContainerKey"<< m_gFexSRJetContainerKey);
  ATH_MSG_DEBUG("m_gFexRhoContainerKey"<< m_gFexRhoContainerKey);
  ATH_MSG_DEBUG("m_gScalarEJwojContainerKey"<< m_gScalarEJwojContainerKey);
  ATH_MSG_DEBUG("m_gMETComponentsJwojContainerKey"<< m_gMETComponentsJwojContainerKey);
  ATH_MSG_DEBUG("m_gMHTComponentsJwojContainerKey"<< m_gMHTComponentsJwojContainerKey);
  ATH_MSG_DEBUG("m_gMSTComponentsJwojContainerKey"<< m_gMSTComponentsJwojContainerKey);

  ATH_CHECK( m_gFexLRJetContainerKey.initialize() );  
  ATH_CHECK( m_gFexSRJetContainerKey.initialize() );  
  ATH_CHECK( m_gFexRhoContainerKey.initialize() );  

  ATH_CHECK( m_gScalarEJwojContainerKey.initialize() );  
  ATH_CHECK( m_gMETComponentsJwojContainerKey.initialize() );  
  ATH_CHECK( m_gMHTComponentsJwojContainerKey.initialize() );  
  ATH_CHECK( m_gMSTComponentsJwojContainerKey.initialize() );  

  ATH_CHECK( m_gMETComponentsNoiseCutContainerKey.initialize() );
  ATH_CHECK( m_gMETComponentsRmsContainerKey.initialize() );
  ATH_CHECK( m_gScalarENoiseCutContainerKey.initialize() );
  ATH_CHECK( m_gScalarERmsContainerKey.initialize() );

  return AthMonitorAlgorithm::initialize();
}

StatusCode GfexMonitorAlgorithm::fillHistograms( const EventContext& ctx ) const {
  ATH_MSG_DEBUG("GfexMonitorAlgorithm::fillHistograms");
  std::vector<std::reference_wrapper<Monitored::IMonitoredVariable>> variables;

  SG::ReadHandle<xAOD::gFexJetRoIContainer> gFexLRJetContainer{m_gFexLRJetContainerKey, ctx};
  if(!gFexLRJetContainer.isValid()){
    ATH_MSG_ERROR("No gFex LR Jet container found in storegate  "<< m_gFexLRJetContainerKey);
    return StatusCode::SUCCESS;
  }
  SG::ReadHandle<xAOD::gFexJetRoIContainer> gFexSRJetContainer{m_gFexSRJetContainerKey, ctx};
  if(!gFexSRJetContainer.isValid()){
    ATH_MSG_ERROR("No gFex SR Jet container found in storegate  "<< m_gFexSRJetContainerKey);
    return StatusCode::SUCCESS;
  }
  SG::ReadHandle<xAOD::gFexJetRoIContainer> gFexRhoContainer{m_gFexRhoContainerKey, ctx};
  if(!gFexRhoContainer.isValid()){
    ATH_MSG_ERROR("No gFex rho container found in storegate  "<< m_gFexRhoContainerKey);
    return StatusCode::SUCCESS;
  }
  SG::ReadHandle<xAOD::gFexGlobalRoIContainer> gScalarEJwojContainer{m_gScalarEJwojContainerKey, ctx};
  if(!gScalarEJwojContainer.isValid()){
    ATH_MSG_ERROR("No scalar energy jets without jets container found in storegate  "<< m_gScalarEJwojContainerKey);
    return StatusCode::SUCCESS;
  }
  SG::ReadHandle<xAOD::gFexGlobalRoIContainer> gMETComponentsJwojContainer{m_gMETComponentsJwojContainerKey, ctx};
  if(!gMETComponentsJwojContainer.isValid()){
    ATH_MSG_ERROR("No MET components for jets without jets container found in storegate  "<< m_gMETComponentsJwojContainerKey);
    return StatusCode::SUCCESS;
  }
  SG::ReadHandle<xAOD::gFexGlobalRoIContainer> gMHTComponentsJwojContainer{m_gMHTComponentsJwojContainerKey, ctx};
  if(!gMHTComponentsJwojContainer.isValid()){
    ATH_MSG_ERROR("No MHT components for jets without jets container found in storegate  "<< m_gMHTComponentsJwojContainerKey);
    return StatusCode::SUCCESS;
  }
  SG::ReadHandle<xAOD::gFexGlobalRoIContainer> gMSTComponentsJwojContainer{m_gMSTComponentsJwojContainerKey, ctx};
  if(!gMSTComponentsJwojContainer.isValid()){
    ATH_MSG_ERROR("No MST components jets without jets container found in storegate  "<< m_gMSTComponentsJwojContainerKey);
    return StatusCode::SUCCESS;
  }
  SG::ReadHandle<xAOD::gFexGlobalRoIContainer> gMETComponentsNoiseCutContainer{m_gMETComponentsNoiseCutContainerKey, ctx};
  if(!gMETComponentsNoiseCutContainer.isValid()){
    ATH_MSG_ERROR("No MET components calculated with Noise Cut Algorithm in storegate  "<< m_gMETComponentsNoiseCutContainerKey);
    return StatusCode::SUCCESS;
  }
  SG::ReadHandle<xAOD::gFexGlobalRoIContainer> gMETComponentsRmsContainer{m_gMETComponentsRmsContainerKey, ctx};
  if(!gMETComponentsRmsContainer.isValid()){
    ATH_MSG_ERROR("No MET components calculated with Rms Algorithm in storegate  "<< m_gMETComponentsRmsContainerKey);
    return StatusCode::SUCCESS;
  }
  SG::ReadHandle<xAOD::gFexGlobalRoIContainer> gScalarENoiseCutContainer{m_gScalarENoiseCutContainerKey, ctx};
  if(!gScalarENoiseCutContainer.isValid()){
    ATH_MSG_ERROR("No scalar energy Noise Cut container found in storegate  "<< m_gScalarENoiseCutContainerKey);
    return StatusCode::SUCCESS;
  }
  SG::ReadHandle<xAOD::gFexGlobalRoIContainer> gScalarERmsContainer{m_gScalarERmsContainerKey, ctx};
  if(!gScalarENoiseCutContainer.isValid()){
    ATH_MSG_ERROR("No scalar energy Noise Cut container found in storegate  "<< m_gScalarENoiseCutContainerKey);
    return StatusCode::SUCCESS;
  }

  // variables for histograms
  auto gtype = Monitored::Scalar<int>("gFexType",0.0);
  auto gFexLRJeteT = Monitored::Scalar<float>("gFexLRJetTransverseEnergy",0.0);
  auto gFexLRJeteta = Monitored::Scalar<float>("gFexLRJetEta",0.0);
  auto gFexLRJetphi = Monitored::Scalar<float>("gFexLRJetPhi",0.0);
  auto gFexSRJeteT = Monitored::Scalar<float>("gFexSRJetTransverseEnergy",0.0);
  auto gFexSRJeteta = Monitored::Scalar<float>("gFexSRJetEta",0.0);
  auto gFexSRJetphi = Monitored::Scalar<float>("gFexSRJetPhi",0.0);
  auto gFexRhoeT = Monitored::Scalar<float>("gFexRhoTransverseEnergy",0.0);
  auto gFexMet = Monitored::Scalar<float>("gFexMET",0.0);
  auto gFexSumEt = Monitored::Scalar<float>("gFexSumET",0.0);
  auto METx = Monitored::Scalar<float>("gFexMETx",0.0);
  auto METy = Monitored::Scalar<float>("gFexMETy",0.0);
  auto MHTx = Monitored::Scalar<float>("gFexMHTx",0.0);
  auto MHTy = Monitored::Scalar<float>("gFexMHTy",0.0);
  auto MSTx = Monitored::Scalar<float>("gFexMSTx",0.0);
  auto MSTy = Monitored::Scalar<float>("gFexMSTy",0.0);
  auto METx_NoiseCut = Monitored::Scalar<float>("gFexMETx_NoiseCut",0.0);
  auto METy_NoiseCut = Monitored::Scalar<float>("gFexMETy_NoiseCut",0.0);
  auto METx_Rms = Monitored::Scalar<float>("gFexMETx_Rms",0.0);
  auto METy_Rms = Monitored::Scalar<float>("gFexMETy_Rms",0.0);
  auto gFexMet_NoiseCut = Monitored::Scalar<float>("gFexMET_NoiseCut",0.0);
  auto gFexSumEt_NoiseCut = Monitored::Scalar<float>("gFexSumET_NoiseCut",0.0);
  auto gFexMet_Rms = Monitored::Scalar<float>("gFexMET_Rms",0.0);
  auto gFexSumEt_Rms = Monitored::Scalar<float>("gFexSumET_Rms",0.0);

  for(const xAOD::gFexJetRoI* gFexLRJetRoI : *gFexLRJetContainer){
    gFexLRJeteT=gFexLRJetRoI->gFexTobEt();
    gFexLRJeteta=gFexLRJetRoI->eta();
    gFexLRJetphi=gFexLRJetRoI->phi();
    gtype=gFexLRJetRoI->gFexType();
    fill(m_packageName,gFexLRJeteta,gFexLRJetphi,gFexLRJeteT,gtype);
  }

  for(const xAOD::gFexJetRoI* gFexSRJetRoI : *gFexSRJetContainer){
    gFexSRJeteT=gFexSRJetRoI->gFexTobEt();
    gFexSRJeteta=gFexSRJetRoI->eta();
    gFexSRJetphi=gFexSRJetRoI->phi();
    gtype=gFexSRJetRoI->gFexType();
    fill(m_packageName,gFexSRJeteta,gFexSRJetphi,gFexSRJeteT,gtype);
  }

  for(const xAOD::gFexJetRoI* gFexRhoRoI : *gFexRhoContainer){
    gFexRhoeT=gFexRhoRoI->gFexTobEt();
    gtype=gFexRhoRoI->gFexType();
    fill(m_packageName,gFexRhoeT, gtype);
  }

  for(const xAOD::gFexGlobalRoI* gScalarEJwoj : *gScalarEJwojContainer){
    gFexSumEt=gScalarEJwoj->METquantityTwo();
    gFexMet=gScalarEJwoj->METquantityOne();
    fill(m_packageName,gFexMet,gFexSumEt);
  }

  for (const xAOD::gFexGlobalRoI* gMETComponentsJwoj : *gMETComponentsJwojContainer) {
    METx=gMETComponentsJwoj->METquantityOne();
    METy=gMETComponentsJwoj->METquantityTwo();
    fill(m_packageName,METx);
    fill(m_packageName,METy);
  }

  for (const xAOD::gFexGlobalRoI* gMHTComponentsJwoj : *gMHTComponentsJwojContainer) {
    MHTx=gMHTComponentsJwoj->METquantityOne();
    MHTy=gMHTComponentsJwoj->METquantityTwo();
    fill(m_packageName,MHTx,MHTy);
  }

  for (const xAOD::gFexGlobalRoI* gMSTComponentsJwoj : *gMSTComponentsJwojContainer) {
    MSTx=gMSTComponentsJwoj->METquantityOne();
    MSTy=gMSTComponentsJwoj->METquantityTwo();
    fill(m_packageName,MSTx,MSTy);
  }

  for (const xAOD::gFexGlobalRoI* gMETComponentsNoiseCut : *gMETComponentsNoiseCutContainer) {
    METx_NoiseCut=gMETComponentsNoiseCut->METquantityOne();
    METy_NoiseCut=gMETComponentsNoiseCut->METquantityTwo();
    fill(m_packageName,METx_NoiseCut,METy_NoiseCut);
  }

  for (const xAOD::gFexGlobalRoI* gMETComponentsRms : *gMETComponentsRmsContainer) {
    METx_Rms=gMETComponentsRms->METquantityOne();
    METy_Rms=gMETComponentsRms->METquantityTwo();
    fill(m_packageName,METx_Rms,METy_Rms);
  }

  for (const xAOD::gFexGlobalRoI* gScalarENoiseCut : *gScalarENoiseCutContainer) {
    gFexMet_NoiseCut=gScalarENoiseCut->METquantityOne();
    gFexSumEt_NoiseCut=gScalarENoiseCut->METquantityTwo();
    fill(m_packageName,gFexMet_NoiseCut,gFexSumEt_NoiseCut);
  }

  for (const xAOD::gFexGlobalRoI* gScalarERms : *gScalarERmsContainer) {
    gFexMet_Rms=gScalarERms->METquantityOne();
    gFexSumEt_Rms=gScalarERms->METquantityTwo();
    fill(m_packageName,gFexMet_Rms,gFexSumEt_Rms);
  }

  fill(m_packageName,variables);
  variables.clear();
  return StatusCode::SUCCESS;
}

